### BC Hydro IPP - Resource Analysis
#### Storage Hydro

BC Hydro acquires power from Independent Power Producers (IPPs) to help meet B.C.'s electricity needs.
IPPs include power production companies, municipalities, First Nations and customers.

The analysis is carried out on the most recent [IPP supply list](https://www.bchydro.com/content/dam/BCHydro/customer-portal/documents/corporate/independent-power-producers-calls-for-power/independent-power-producers/ipp-supply-list-in-operation.pdf) and [IPP supply map](https://www.bchydro.com/content/dam/BCHydro/customer-portal/documents/corporate/independent-power-producers-calls-for-power/independent-power-producers/ipp-supply-map.pdf) obtained from [BC Hydro Website](https://www.bchydro.com/work-with-us/selling-clean-energy/meeting-energy-needs/how-power-is-acquired.html).

In [1]:
import pandas as pd
import altair as alt

In [2]:
data = pd.read_csv('IPP__REGION_list.csv')

#### Initial Description of Data


As of April 1, 2025, BC Hydro has 119 Electricity Purchase Agreements (EPAs) with IPPs whose projects achieved commercial operation, as defined in their EPA and are considered to be in operation. In total, these projects are capable of delivering approximately 18,662 gigawatt hours of annual supply and approximately 5,298 megawatts of capacity.

In [3]:
len(data)

119

In [4]:
IPP_Capacity = round(data['Capacity(MW)'].sum(),2)
IPP_Capacity

5298.84

In [5]:
IPP_Energy = round(data['Energy(GWh/yr)'].sum(),2)
IPP_Energy

18662.3

In [6]:
data.replace({'Sub_region':{'CI':'Central Interior',
                               'LM':'Lower Mainland',
                               'VI':'Vancouver Island',
                               'NC':'North Coast',
                               'EK':'East Kootenay',
                               'PR':'Peace River',
                               'KLN':'Kelly Lake Nicola',
                               'RSAC':'Revelstoke Ashton Creek',
                               'SK':'Selkirk'}}, inplace = True)

In [7]:
peace_list = ['Central Interior','Peace River']

In [8]:
columbia_list = ['Revelstoke Ashton Creek', 'Kelly Lake Nicola', 'Mica', 'Selkirk', 'East Kootenay']

In [9]:
data.loc[data['Sub_region'].isin(columbia_list),'Region'] = 'Columbia Region'
data.loc[data['Sub_region'].isin(peace_list),'Region'] = 'Peace Region'
data.loc[data['Sub_region'] == 'Lower Mainland','Region'] = 'Lower Mainland'
data.loc[data['Sub_region'] == 'North Coast','Region'] = 'North Coast'
data.loc[data['Sub_region'] == 'Vancouver Island','Region'] = 'Vancouver Island'

In [10]:
columns = data.columns.to_list()
for column in columns:
    if data[column].dtype == 'object':
        data[column] = data[column].str.strip()

The capacity and GWhr contribution of various resources by Type is given by the data_capacity_type and data_energy_type respectively 

In [11]:
data_capacity_type = data.groupby('Type').agg({'Capacity(MW)':'sum'}).sort_values(by = 'Capacity(MW)',ascending = False).reset_index()
data_energy_type = data.groupby('Type').agg({'Energy(GWh/yr)':'sum'}).sort_values(by = 'Energy(GWh/yr)',ascending = False).reset_index()

In [12]:
data_energy_type

Type  Energy(GWh/yr)
0  Non-Storage Hydro          7336.4
1      Storage Hydro          4933.3
2            Biomass          2977.6
3               Wind          2161.8
4  Gas Fired Thermal           904.7
5                MSW           166.0
6                ERG           141.0
7             Biogas            37.6
8              Solar             3.9

In [13]:
data_capacity_type

Type  Capacity(MW)
0  Non-Storage Hydro      1983.200
1      Storage Hydro      1292.100
2            Biomass       817.900
3               Wind       747.000
4  Gas Fired Thermal       403.800
5                MSW        24.800
6                ERG        22.800
7             Biogas         5.155
8              Solar         2.090

Storage Hydro IPP projects are capable of delivering approximately 4933 gigawatt hours of annual supply and approximately 1292 megawatts of capacity

#### Storage Hydro 
Storage hydro resources are hydroelectric facilities that generate power relying on stored water in reservoirs. Most hydroelectric power comes from the potential energy of dammed water driving a water turbine and generator.([source](https://en.wikipedia.org/wiki/Hydroelectricity#cite_note-25:~:text=Most%20hydroelectric%20power,%5B25%5D)).

BC Hydro has Electricity Purchase Agreement for Storage Hydro with 14 Independent Power Producers. In total, these 14 projects are capable of delivering approximately 4933 gigawatt hours of annual supply and approximately 1292 megawatts of capacity. The capacity and GWhr contribution of Storage Hydro in various regions are given by the dataframes energy_SH, capacity_SH respectively. 

In [14]:
data_SH = data[data['Type'] == 'Storage Hydro'].reset_index(drop=True)
energy_SH = data_SH.groupby('Region').agg({'Energy(GWh/yr)':'sum'}).sort_values(by = 'Energy(GWh/yr)',ascending = False).reset_index()
capacity_SH = data_SH.groupby('Region').agg({'Capacity(MW)':'sum'}).sort_values(by = 'Capacity(MW)',ascending = False).reset_index()

In [15]:
len(data_SH)

14

In [16]:
energy_SH

Region  Energy(GWh/yr)  Capacity(MW)
0       North Coast          3546.2         945.2
1   Columbia Region          1199.7         306.1
2  Vancouver Island           122.6          29.9
3    Lower Mainland            64.8          10.9

In [17]:
capacity_SH

Region  Capacity(MW)
0       North Coast         945.2
1   Columbia Region         306.1
2  Vancouver Island          29.9
3    Lower Mainland          10.9

For Storage Hydro, the capacity and energy contributions of each region individually analysed 

#### 1. Lower Mainland


For the lower mainland region, the capacity and GWhr contribution of Storage Hydro resource by location is given in the capacity_LM_Loc and energy_LM_Loc dataframes respectively.


In [40]:
data_LM = data[(data['Type'] == 'Storage Hydro') & (data['Region']=='Lower Mainland')].reset_index(drop=True)
energy_LM_Loc = data_LM.groupby('Location').agg({'Energy(GWh/yr)':'sum',
                                          'Capacity(MW)':'sum'}).sort_values(by = 'Energy(GWh/yr)',ascending = False).reset_index()
capacity_LM_Loc = data_LM.groupby('Location').agg({'Capacity(MW)':'sum'}).sort_values(by = 'Capacity(MW)',ascending = False).reset_index()

In [41]:
capacity_LM_Loc

Location  Capacity(MW)
0   Sechelt           9.3
1  Squamish           1.6

In [42]:
energy_LM_Loc

Location  Energy(GWh/yr)  Capacity(MW)
0   Sechelt            53.6           9.3
1  Squamish            11.2           1.6

The bar chart energy_LM_Loc_chart, shows for Storage Hydro, the energy contribution of the project and the call process by which it was acquired. 

The bar chart capacity_LM_Loc_chart, shows for Storage Hydro, the capacity contribution of each of these projects and the call process by which it was acquired.


In [21]:
title = alt.TitleParams('The Annual GWhr contribution of Storage Hydro in Lower Mainland by IPP Projects')

capacity_IPP = alt.Chart(data_LM, title = title).mark_bar().encode(
                    alt.Y('Project_name', title='Project Name',sort='x'),
                    alt.X('sum(Energy(GWh/yr))', title='Energy(GWh/yr)'),
                    alt.Color('Call_process', title = 'Call Process')).properties( width=300)
energy_LM_Loc_chart = capacity_IPP + capacity_IPP.mark_text(
    color = 'black', align='left', dx = 3).encode(text='sum(Energy(GWh/yr))',color=alt.value('black'))
energy_LM_Loc_chart

alt.LayerChart(...)

In [22]:
title = alt.TitleParams('The Capacity contribution of Storage Hydro in Lower Mainland by IPP Projects')

capacity_IPP = alt.Chart(data_LM, title = title).mark_bar().encode(
                    alt.Y('Project_name', title='Project Name',sort='x'),
                    alt.X('sum(Capacity(MW))', title='Capacity(MW)'),
                    alt.Color('Call_process', title = 'Call Process')).properties( width=300)
cap_LM_Loc_chart = capacity_IPP + capacity_IPP.mark_text(
    color = 'black', align='left', dx = 3).encode(text='sum(Capacity(MW))',color=alt.value('black'))
cap_LM_Loc_chart

alt.LayerChart(...)

#### 2. Vancouver Island


For the Vancouver Island region, the capacity and GWhr contribution of the Storage Hydro resource by location is given in the capacity_VI_Loc and energy_VI_Loc dataframes respectively.

In [43]:
data_VI = data[(data['Type'] == 'Storage Hydro') & (data['Region']=='Vancouver Island')].reset_index(drop=True)
energy_VI_Loc = data_VI.groupby('Location').agg({'Energy(GWh/yr)':'sum',
                                                'Capacity(MW)':'sum'}).sort_values(by = 'Energy(GWh/yr)',ascending = False).reset_index()
capacity_VI_Loc = data_VI.groupby('Location').agg({'Capacity(MW)':'sum'}).sort_values(by = 'Capacity(MW)',ascending = False).reset_index()

In [44]:
energy_VI_Loc

Location  Energy(GWh/yr)  Capacity(MW)
0    Zeballos            92.6          21.9
1  Port Alice            30.0           8.0

In [25]:
capacity_VI_Loc

Location  Capacity(MW)
0    Zeballos          21.9
1  Port Alice           8.0

The bar chart energy_VI_Loc_chart, shows the energy contribution for each of these Storage Hydro projects and the call process by which it was acquired. 

The bar chart capacity_VI_Loc_chart, shows the capacity contribution for each of these Storage Hydro projects and the call process by which it was acquired.


In [26]:
title = alt.TitleParams('The Annual GWhr contribution of Storage Hydro in Vancouver Island by IPP Projects')

capacity_IPP = alt.Chart(data_VI, title = title).mark_bar().encode(
                    alt.Y('Project_name', title='Project Name',sort='x'),
                    alt.X('sum(Energy(GWh/yr))', title='Energy(GWh/yr)'),
                    alt.Color('Call_process', title = 'Call Process')).properties( width=300)
energy_VI_Loc_chart = capacity_IPP + capacity_IPP.mark_text(
    color = 'black', align='left', dx = 3).encode(text='sum(Energy(GWh/yr))',color=alt.value('black'))
energy_VI_Loc_chart

alt.LayerChart(...)

In [27]:
title = alt.TitleParams('The Capacity contribution of Storage Hydro in Vancouver Island by IPP Projects')

capacity_IPP = alt.Chart(data_VI, title = title).mark_bar().encode(
                    alt.Y('Project_name', title='Project Name',sort='x'),
                    alt.X('sum(Capacity(MW))', title='Capacity(MW)'),
                    alt.Color('Call_process', title = 'Call Process')).properties( width=300)
cap_VI_Loc_chart = capacity_IPP + capacity_IPP.mark_text(
    color = 'black', align='left', dx = 3).encode(text='sum(Capacity(MW))',color=alt.value('black'))
cap_VI_Loc_chart

alt.LayerChart(...)

#### 3. Peace Region

For Peace region, there are no Storage Hydro projects in operation right now

In [28]:
data_PR = data[(data['Type'] == 'Storage Hydro') & (data['Region']=='Peace Region')].reset_index(drop=True)

In [29]:
len(data_PR)

0

#### 4. Columbia Region

For Columbia region, the capacity and GWhr contribution of the Storage Hydro resource by location is given in the capacity_CR_Loc and energy_CR_Loc dataframes respectively.

In [45]:
data_CR = data[(data['Type'] == 'Storage Hydro') & (data['Region']=='Columbia Region')].reset_index(drop=True)
energy_CR_Loc = data_CR.groupby('Location').agg({'Energy(GWh/yr)':'sum','Capacity(MW)':'sum'}).sort_values(by = 'Energy(GWh/yr)',ascending = False).reset_index()
capacity_CR_Loc = data_CR.groupby('Location').agg({'Capacity(MW)':'sum'}).sort_values(by = 'Capacity(MW)',ascending = False).reset_index()

In [46]:
energy_CR_Loc

Location  Energy(GWh/yr)  Capacity(MW)
0     Slocan           767.0         185.0
1  Castlegar           428.7         120.0
2    Kelowna             4.0           1.1

In [32]:
capacity_CR_Loc

Location  Capacity(MW)
0     Slocan         185.0
1  Castlegar         120.0
2    Kelowna           1.1

The bar chart energy_CR_Loc_chart, shows the energy contribution of the project and the call process by which it was acquired. 

The bar chart capacity_CR_Loc_chart, shows the capacity contribution of each of these projects and the call process by which it was acquired.

In [33]:
title = alt.TitleParams('The Annual GWhr contribution of Storage Hydro in Columbia Region by IPP Project')

capacity_IPP = alt.Chart(data_CR, title = title).mark_bar().encode(
                    alt.Y('Project_name', title='Project Name',sort='x'),
                    alt.X('sum(Energy(GWh/yr))', title='Energy(GWh/yr)'),
                    alt.Color('Call_process', title = 'Call Process')).properties( width=300)
energy_CR_Loc_chart = capacity_IPP + capacity_IPP.mark_text(
    color = 'black', align='left', dx = 3).encode(text='sum(Energy(GWh/yr))',color=alt.value('black'))
energy_CR_Loc_chart

alt.LayerChart(...)

In [34]:
title = alt.TitleParams('The Capacity contribution of Storage Hydro in Columbia Region by IPP Project')

capacity_IPP = alt.Chart(data_CR, title = title).mark_bar().encode(
                    alt.Y('Project_name', title='Project Name',sort='x'),
                    alt.X('sum(Capacity(MW))', title='Capacity(MW)'),
                    alt.Color('Call_process', title = 'Call Process')).properties( width=300)
cap_CR_Loc_chart = capacity_IPP + capacity_IPP.mark_text(
    color = 'black', align='left', dx = 3).encode(text='sum(Capacity(MW))',color=alt.value('black'))
cap_CR_Loc_chart

alt.LayerChart(...)

#### 5. North Coast

For North Coast, the capacity and GWhr contribution of the Storage Hydro resource by location is given in the capacity_NC_Loc and energy_NC_Loc dataframes respectively.

In [47]:
data_NC = data[(data['Type'] == 'Storage Hydro') & (data['Region']=='North Coast')].reset_index(drop=True)
energy_NC_Loc = data_NC.groupby('Location').agg({'Energy(GWh/yr)':'sum',
                                                 'Capacity(MW)':'sum'
                                                }).sort_values(by = 'Energy(GWh/yr)',ascending = False).reset_index()
capacity_NC_Loc = data_NC.groupby('Location').agg({'Capacity(MW)':'sum'}).sort_values(by = 'Capacity(MW)',ascending = False).reset_index()

In [48]:
energy_NC_Loc

Location  Energy(GWh/yr)  Capacity(MW)
0        Kitimat          3307.0         896.0
1        Stewart           152.6          31.0
2  Prince Rupert            57.0           7.2
3       Sandspit            20.0           6.0
4     Dease Lake             5.0           3.0
5          Atlin             4.6           2.0

In [37]:
capacity_NC_Loc

Location  Capacity(MW)
0        Kitimat         896.0
1        Stewart          31.0
2  Prince Rupert           7.2
3       Sandspit           6.0
4     Dease Lake           3.0
5          Atlin           2.0

In [38]:
title = alt.TitleParams('The Annual GWhr contribution of Storage Hydro in North Coast by IPP Project')

capacity_IPP = alt.Chart(data_NC, title = title).mark_bar().encode(
                    alt.Y('Project_name', title='Project Name',sort='x'),
                    alt.X('sum(Energy(GWh/yr))', title='Energy(GWh/yr)'),
                    alt.Color('Call_process', title = 'Call Process')).properties( width=300)
energy_NC_Loc_chart = capacity_IPP + capacity_IPP.mark_text(
    color = 'black', align='left', dx = 3).encode(text='sum(Energy(GWh/yr))',color=alt.value('black'))
energy_NC_Loc_chart

alt.LayerChart(...)

In [39]:
title = alt.TitleParams('The Capacity contribution of Storage Hydro in North Coast by IPP Project')

capacity_IPP = alt.Chart(data_NC, title = title).mark_bar().encode(
                    alt.Y('Project_name', title='Project Name',sort='x'),
                    alt.X('sum(Capacity(MW))', title='Capacity(MW)'),
                    alt.Color('Call_process', title = 'Call Process')).properties( width=300)
cap_NC_Loc_chart = capacity_IPP + capacity_IPP.mark_text(
    color = 'black', align='left', dx = 3).encode(text='sum(Capacity(MW))',color=alt.value('black'))
cap_NC_Loc_chart

alt.LayerChart(...)

#### Conclusion

This study was cariied out to learn about the capacity and energy distribution of Storage Hydro IPP resources around various regions in the province. 